src/mame/drivers/firebeat.cpp

```diff
@@ -649,170 +726,361 @@ void firebeat_state::extend_board_irq_w(offs_t offset, uint32_t data, uint32_t m
 
 /*****************************************************************************/
 
-void firebeat_state::lamp_output_w(offs_t offset, uint32_t data, uint32_t mem_mask)
+uint32_t firebeat_state::input_r(offs_t offset, uint32_t mem_mask)
 {
-	// -------- -------- -------- xxxxxxxx   Status LEDs (active low)
+	uint32_t r = 0;
+
+	if (ACCESSING_BITS_24_31)
+	{
+		r |= (ioport("IN0")->read() & 0xff) << 24;
```

Comment:
You made a lot of comments about this kind of thing in this PR, and I remember one or two other comments in other PRs. Am I right in assuming that it will always be preferable to get rid of the ioport lookups, especially in what would be considered hot paths?

I'll be more proactive about looking for that in drivers I'm working on.

Reply:
runtime tag looks are very slow to the point I've floated the idea of aborting when they happen, so they get weeded out, so yeah, getting rid of them is always preferable

Reply:
They're not too bad if they only happen at reset time – then you’ll get an error straight away on starting the machine if there’s an issue, and it won’t be slowing you down much.  But if you do them any other time, they’re slow, and if you can use a `required_ioport` you’ll get an error if there’s something wrong without having to hit the code path that actually uses it.

So yeah, avoid tag lookups on the hot path.

Reply:
Sounds reasonable to me. I'll keep it in mind from now on. There shouldn't be any more such lookups in the Firebeat driver anymore after my latest commits. I'll try giving ksys573 a once over the next time I work on it. Both drivers are fairly old so I'm sure some more modernization could be done.